relay-profiling/src/sample.rs

```diff
@@ -0,0 +1,429 @@
+                    && self.device.locale.is_some()
+                    && self.device.manufacturer.is_some()
+                    && self.device.model.is_some()
+            }
+            Platform::Python => self.runtime.is_some(),
+            Platform::Node => self.runtime.is_some(),
+            _ => true,
+        }
+    }
+
+    /// Removes a sample when it's the only sample on its thread
+    fn remove_single_samples_per_thread(&mut self) {
+        let mut sample_count_by_thread_id: HashMap<u64, u32> = HashMap::new();
+
+        for sample in &self.profile.samples {
+            *sample_count_by_thread_id
+                .entry(sample.thread_id)
+                .or_default() += 1;
+        }
+
+        // Only keep data from threads with more than 1 sample so we can calculate a duration
+        sample_count_by_thread_id.retain(|_, count| *count > 1);
+
+        self.profile
+            .samples
+            .retain(|sample| sample_count_by_thread_id.contains_key(&sample.thread_id));
+    }
+
+    fn strip_pointer_authentication_code(&mut self) {
+        self.profile
+            .strip_pointer_authentication_code(&self.platform, &self.device.architecture);
+    }
+}
+
+pub fn expand_sample_profile(payload: &[u8]) -> Result<Vec<Vec<u8>>, ProfileError> {
+    let profile = parse_profile(payload)?;
+
+    if !profile.valid() {
+        return Err(ProfileError::MissingProfileMetadata);
+    }
+
+    let mut items: Vec<Vec<u8>> = Vec::new();
+
+    for transaction in &profile.transactions {
+        let mut new_profile = profile.clone();
+
+        new_profile.transactions.clear();
+        new_profile.transactions.push(transaction.clone());
+
+        new_profile.profile.samples.retain_mut(|sample| {
+            if transaction.relative_start_ns <= sample.relative_timestamp_ns
+                && sample.relative_timestamp_ns <= transaction.relative_end_ns
+            {
+                sample.relative_timestamp_ns -= transaction.relative_start_ns;
+                true
+            } else {
+                false
+            }
+        });
+
+        match serde_json::to_vec(&new_profile) {
+            Ok(payload) => items.push(payload),
+            Err(_) => {
+                return Err(ProfileError::CannotSerializePayload);
+            }
+        };
+    }
+
+    Ok(items)
+}
+
+fn parse_profile(payload: &[u8]) -> Result<SampleProfile, ProfileError> {
+    let mut profile: SampleProfile =
+        serde_json::from_slice(payload).map_err(ProfileError::InvalidJson)?;
+
+    profile.remove_single_samples_per_thread();
+
+    if profile.profile.samples.is_empty() {
+        return Err(ProfileError::NotEnoughSamples);
+    }
+
+    profile
+        .transactions
+        .retain(|transaction| transaction.valid());
+
+    if profile.transactions.is_empty() {
+        return Err(ProfileError::NoTransactionAssociated);
+    }
+
+    profile.profile.frames.retain(|frame| frame.valid());
```

Comment:
Could `frames` be empty after this? If so, should an error be returned in that case?

Reply:
Actually, since frames are indexed (and stacks as well), I can't remove any of them, valid or invalid. I'll remove this.